photon-ml/src/test/scala/com/linkedin/photon/ml/algorithm/CoordinateDescentTest.scala

```diff
@@ -76,19 +99,33 @@ class CoordinateDescentTest {
       doReturn(score).when(coordinate).score(model)
       doReturn((model, tracker)).when(coordinate).updateModel(model)
       doReturn((model, tracker)).when(coordinate).updateModel(model, score)
+
+      // Validation evaluator mock setup
+      doReturn(validationScore).when(validationScore). +(validationScore)
+      doReturn(modelScores).when(validationScore).scores
+      doReturn(validationScore).when(model).score(validationData)
+      doReturn(validationScore).when(validationScore).setName(Matchers.anyString())
+      doReturn(validationScore).when(validationScore).persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
     }
 
     // Run coordinate descent
-    val coordinateDescent = new CoordinateDescent(coordinates, evaluator, None, logger)
-    coordinateDescent.run(numIterations, gameModel)
+    val validationDataAndEvaluators = if (validationEvaluators.isEmpty) None else Option(validationData, validationEvaluators)
+    val coordinateDescent = new CoordinateDescent(coordinates, lossEvaluator, validationDataAndEvaluators, logger)
+    coordinateDescent.run(iterationCount, gameModel)
 
     // Verify the calls to updateModel
     if (coordinates.length == 1) {
-      verify(coordinates.head._2, times(numIterations)).updateModel(models.head)
-
+      verify(coordinates.head._2, times(iterationCount)).updateModel(models.head)
     } else {
       (coordinates zip models).map { case ((coordinateId, coordinate), model) =>
-        verify(coordinate, times(numIterations)).updateModel(model, score)
+        verify(coordinate, times(iterationCount)).updateModel(model, score)
+      }
+    }
+
+    // Verify the calls to the validation evaluator(s), if any
+    validationDataAndEvaluators.map { case (data, evaluators) =>
+      evaluators.map { case (evaluator) =>
+        verify(evaluator, times(iterationCount * coordinateCount)).evaluate(modelScores)
```

Comment:
I think that this test should be mostly reverted and the new changes tested in a new test method. The new test method would mock `Evaluator.evaluate` to return different metrics depending on the input mocked KeyValueScore objects.

I'm thinking 3 iterations, and 4 mocked KeyValueScores objects (base + one per iteration) that each point to the next in the chain. In addition, 2 mocked Evaluators: the first ranks the second model the highest, and the second ranks the first model the highest. We should check that we get back the second model, as either the first or third would be incorrect (first means we're using wrong Evaluator, third means we're not using either Evaluator).